src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs

```diff
@@ -0,0 +1,348 @@
+            if (!syntaxFacts.IsStringLiteral(token))
+            {
+                return true;
+            }
+
+            if (!IsMethodOrConstructorArgument(token, syntaxFacts) && 
+                !HasRegexLanguageComment(token, syntaxFacts, out _))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool HasRegexLanguageComment(
+            SyntaxToken token, ISyntaxFactsService syntaxFacts, out RegexOptions options)
+        {
+            if (HasRegexLanguageComment(token.GetPreviousToken().TrailingTrivia, syntaxFacts, out options))
+            {
+                return true;
+            }
+
+            for (var node = token.Parent; node != null; node = node.Parent)
+            {
+                if (HasRegexLanguageComment(node.GetLeadingTrivia(), syntaxFacts, out options))
+                {
+                    return true;
+                }
+            }
+
+            options = default;
+            return false;
+        }
+
+        private static bool HasRegexLanguageComment(
+            SyntaxTriviaList list, ISyntaxFactsService syntaxFacts, out RegexOptions options)
+        {
+            foreach (var trivia in list)
+            {
+                if (HasRegexLanguageComment(trivia, syntaxFacts, out options))
+                {
+                    return true;
+                }
+            }
+
+            options = default;
+            return false;
+        }
+
+        private static bool HasRegexLanguageComment(
+            SyntaxTrivia trivia, ISyntaxFactsService syntaxFacts, out RegexOptions options)
+        {
+            if (syntaxFacts.IsRegularComment(trivia))
+            {
+                var text = trivia.ToString();
+                var match = s_languageCommentDetector.Match(text);
+                if (match.Success)
+                {
+                    options = RegexOptions.None;
+
+                    var optionGroup = match.Groups["option"];
+                    foreach (Capture capture in optionGroup.Captures)
+                    {
+                        if (s_nameToOption.TryGetValue(capture.Value, out var specificOption))
+                        {
+                            options |= specificOption;
+                        }
+                        else
+                        {
+                            break;
+                        }
+                    }
+
+                    return true;
+                }
+            }
+
+            options = default;
+            return false;
+        }
+
+        private static bool IsMethodOrConstructorArgument(SyntaxToken token, ISyntaxFactsService syntaxFacts)
+            => syntaxFacts.IsLiteralExpression(token.Parent) &&
+               syntaxFacts.IsArgument(token.Parent.Parent);
+
+        private static HashSet<string> GetMethodNamesOfInterest(INamedTypeSymbol regexType, ISyntaxFactsService syntaxFacts)
```

Comment:
>GetMethodNamesOfInterest [](start = 39, length = 24)

I didn't understand the concept of "method names of interest". I expected the only method of interest where the `Regex` constructors.
Also, I worry that the parameter name "pattern" may not be a robust detection mechanism (did we check corefx, mono, etc)?

Reply:
I will comment.  But there are far more than the regex constructors.  For example:

![image](https://user-images.githubusercontent.com/4564579/44162676-daf66a00-a075-11e8-8ae5-b6c97954b219.png)

> Also, I worry that the parameter name "pattern" may not be a robust detection mechanism (did we check corefx, mono, etc)?

I checked the Regex class in .Net and CoreFx.  This was a suitable 'pattern' for finding applicable locations.  I didn't check mono.  But if their regex class doesn't match teh same parameter names as .Net that would be bad as well given that named-parameters are a thing :)

Reply:
Doc'ed

Reply:
Or, put another way, inside the Regex class, a parameter called 'pattern' is always a pattern.  Parameters not called pattern are *not* patterns.  That held true in the entire Regex type in standard and core.  

Note: we're only doing detection when using the Regex type itself, not any other type.  For all other cases, you need to use the `/*lang=regex*/` form.

Reply:
Ah. Makes sense.

---
In reply to: [210344510](https://github.com/dotnet/roslyn/pull/25517#discussion_r210344510) [](ancestors = 210344510)